modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/row/SqlRowHandlerTest.java

```diff
@@ -0,0 +1,214 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * Tests for {@link SqlRowHandler}.
+ */
+public class SqlRowHandlerTest extends IgniteAbstractTest {
+    private static final RowHandler<RowWrapper> handler = SqlRowHandler.INSTANCE;
+
+    private final long seed = ThreadLocalRandom.current().nextLong();
+
+    private final Random rnd = new Random(seed);
+
+    @BeforeEach
+    void printSeed() {
+        log.info("Using seed: " + seed);
+    }
+
+    @Test
+    public void testBytebufferSerialization() {
+        List<ColumnType> columnTypes = columnTypes();
+        Object[] sourceData = values(columnTypes);
+        RowSchema schema = rowSchema(columnTypes, sourceData);
+
+        int elementsCount = schema.fields().size();
+
+        RowFactory<RowWrapper> factory = handler.factory(schema);
+        RowWrapper src = factory.create(sourceData);
+
+        // Serialization to binary tuple representation.
+        ByteBuffer buf = handler.toByteBuffer(src);
+
+        BinaryTuple tuple = new BinaryTuple(elementsCount, buf);
+        RowWrapper destWrap = factory.wrap(tuple);
+        RowWrapper dest = factory.create(buf);
+
+        for (int i = 0; i < elementsCount; i++) {
+            String msg = schema.fields().get(i).toString();
+
+            assertThat(msg, handler.get(i, src), equalTo(sourceData[i]));
+            assertThat(msg, schema.value(i, tuple), equalTo(sourceData[i]));
+
+            // Binary tuple wrapper must return data in internal format.
+            Object expected = TypeUtils.toInternal(sourceData[i]);
+
+            assertThat(msg, handler.get(i, dest), equalTo(expected));
+            assertThat(msg, handler.get(i, destWrap), equalTo(expected));
+        }
+    }
+
+    @ParameterizedTest(name = "{0} - {1}")
+    @MethodSource("concatTestArguments")
+    public void testConcat(boolean leftTupleRequired, boolean rightTupleRequired) {
+        ConcatTestParameters params = new ConcatTestParameters(leftTupleRequired, rightTupleRequired);
+
+        RowWrapper concatenated = handler.concat(params.left, params.right);
+
+        int leftLen = params.leftData.length;
+        int rightLen = params.rightData.length;
+        int totalElementsCount = leftLen + rightLen;
+
+        assertThat(handler.columnCount(concatenated), equalTo(totalElementsCount));
+
+        // Build combined schema.
+        Builder builder = RowSchema.builder();
+        params.leftSchema.fields().forEach(builder::addField);
+        params.rightSchema.fields().forEach(builder::addField);
+
+        RowSchema concatenatedSchema = builder.build();
+
+        // Serialize.
+        ByteBuffer buf = handler.toByteBuffer(concatenated);
+
+        // Wrap into row.
+        RowWrapper wrapped = handler.factory(concatenatedSchema).wrap(new BinaryTuple(totalElementsCount, buf));
+
+        for (int i = 0; i < Math.max(leftLen, rightLen); i++) {
+            if (i < leftLen) {
```

Comment:
I think is code too complicate for what is supposed to do - concat(row1+row2) = row(cols::row1, cols::row2)
We can just concatenate leftData with rightData and check elements of concatenated row one by one.

Reply:
Split into two cycles.